### PR TITLE
Long running API for dump and restore refs

### PR DESCRIPTION
**Pull request by nopcoder**
Close https://github.com/treeverse/lakeFS/issues/6527

- [ ] add test for dump/restore refs

### CODE REVIEW

File: docs/assets/js/swagger.yml [hunk 1]
```diff
@@ -594,12 +594,62 @@ components:
...
+    RefsRestoreTask:
+      type: object
+      required:
+        - id
+      properties:
+        id:
+          type: string
+          description: ID of the task
+        refs:
+          $ref: "#/components/schemas/RefsDump"
```
Comment (johnnyaug) [inline docs/assets/js/swagger.yml:631]: not used?

File: api/swagger.yml [hunk 2]
```diff
@@ -594,12 +594,62 @@ components:
...
+          type: string
+
     RefsDump:
       type: object
       required:
-        - branches_meta_range_id
-        - tags_meta_range_id
         - commits_meta_range_id
+        - tags_meta_range_id
+        - branches_meta_range_id
```
Comment (johnnyaug) [inline api/swagger.yml:652]: Is the order change significant?

File: cmd/lakectl/cmd/refs_restore.go [hunk 3]
```diff
@@ -45,9 +46,50 @@ Since a bare repo is expected, in case of transient failure, delete the reposito
...
+		DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusAccepted)
+		if resp.JSON202 == nil {
+			Die("Bad response from server", 1)
+		}
+		taskID := resp.JSON202.Id
+		logging.FromContext(ctx).
+			WithField("task_id", taskID).
+			Debug("Submitted refs restore")
+
+		const checkInterval = 3 * time.Second
```
Comment (johnnyaug) [inline cmd/lakectl/cmd/refs_restore.go:64]: This one is not customizable? (I don't really mind either way)

File: pkg/catalog/catalog.go [hunk 4]
```diff
@@ -1748,67 +1753,213 @@ func (c *Catalog) FindMergeBase(ctx context.Context, repositoryID string, destin
...
+				WithError(err).
+				WithFields(logging.Fields{"task_id": taskID, "repository": repository.RepositoryID}).
+				Error("failed to update task status")
+		}
+
+		// make sure we stop based on task completed status, as we may fail
+		if taskStatus.Task.Completed {
+			break
+		}
+	}
```
Comment (johnnyaug) [inline pkg/catalog/catalog.go:1851]: This is duplicate code - can we extract it?

File: pkg/catalog/catalog.go [hunk 5]
```diff
@@ -1748,67 +1753,213 @@ func (c *Catalog) FindMergeBase(ctx context.Context, repositoryID string, destin
...
+			c.log(ctx).
+				WithError(err).
+				WithFields(logging.Fields{"task_id": taskID, "repository": repository.RepositoryID}).
+				Error("failed to %s", step.name)
+			taskStatus.Task.Error = err.Error()
+			taskStatus.Task.Completed = true
+		} else if stepIdx == len(steps)-1 {
+			taskStatus.Task.Completed = true
+		}
+		taskStatus.Task.UpdatedAt = timestamppb.Now()
```
Comment (johnnyaug) [inline pkg/catalog/catalog.go:1934]: Move all assignments of "UpdatedAt" into the updateTaskStatus function?

File: pkg/catalog/catalog.proto [hunk 6]
```diff
@@ -22,3 +22,37 @@ message Entry {
...
 }
+
+// Task is used as field 1 in all status messages
+message Task {
+	string id = 1;
+	bool completed = 2;
+	google.protobuf.Timestamp updated_at = 3;
+	int64 progress = 4;
+	//  string metarange_id = 5;
+	//  CommitData commit = 6;
```
Comment (johnnyaug) [inline pkg/catalog/catalog.proto:33]: Remove?

**Review by johnnyaug — CHANGES_REQUESTED** [review]

File: api/swagger.yml (hunk 2 above)
Comment (nopcoder) [reply api/swagger.yml:652]: nope - it should not have an any effect. also the type name was changed as it is a new operation.

File: cmd/lakectl/cmd/refs_restore.go (hunk 3 above)
Comment (nopcoder) [reply cmd/lakectl/cmd/refs_restore.go:64]: Adding flag to customize interval check also for restore. Default move to common code.

File: pkg/catalog/catalog.go (hunk 4 above)
Comment (nopcoder) [reply pkg/catalog/catalog.go:1851]: will use backoff common code to run the loop - but each one perform different call with different output type

File: pkg/catalog/catalog.go (hunk 5 above)
Comment (nopcoder) [reply pkg/catalog/catalog.go:1934]: Didn't want the function that update the kv to update my message - let me know if you think we should update the function purpose.

**Comment by nopcoder** [comment]
@johnnyaug thanks! I've fixed all the comments and found more issues :(
I'll address the additional issues found and re-request.

**Comment by nopcoder** [comment]
Closing this one - reimplemented this PR and kept the old one as internal before removing compatability.

**Pull request closed without merging**